### ISSUE TITLE
Connection References fail to set on Custom Connections

### ISSUE THREAD

**Issue by tdashworth** [Capgemini/powerapps-packagedeployer-template#55]
## Describe the bug
Within our solution, we have a custom connection, a connection reference to it, and flows that use the connection reference. 

For the deployment, we:
- first deployed with no connection reference config (this failed to activate the Flows)
- now the Solution was installed, the connection could be created
- next the connection id was set via environment variables
- the deployment failed to link the reference to the connection.

```
PackageDeployer Error: 2 : Message: Request failed with: NotFound and error: {"error":{"code":"ConnectionNotFound","message":"Could not find connection '[redacted]' for API '[redacted]'."}} and request url https://api.powerapps.com/providers/Microsoft.PowerApps/scopes/service/apis/[redacted]/connections/[redacted]?api-version=2018-10-01&$expand=permissions($filter=maxAssignedTo('[redacted]')&$filter=environment+eq+'[redacted]' 
Source	: Not Provided
Method	: Not Provided
Date	: 7:26:32 PM
Time	: 4/21/2021
Error	: Request failed with: NotFound and error: {"error":{"code":"ConnectionNotFound","message":"Could not find connection '[redacted]' for API '[redacted]'."}} and request url https://api.powerapps.com/providers/Microsoft.PowerApps/scopes/service/apis/[redacted]/connections/[redacted]?api-version=2018-10-01&$expand=permissions($filter=maxAssignedTo('[redacted]')&$filter=environment+eq+'[redacted]'
Stack Trace	: Not Provided
```

It appears the imported custom connection is created new and therefore the "API" could not be found which explains this screenshot showing the connection reference with a blank "Connection". Looking at the URLs of the connection in Dev and CI (where the issue is) the API names are different. 

![image](https://user-images.githubusercontent.com/9384593/115614377-8f47f400-a2e5-11eb-8b1e-36a8f903bf7a.png)

## To reproduce

Create a solution with a custom connection, and a flow that uses it and therefore a connection reference to it. 

## Expected behaviour
The custom connection reference can be set like any other connection.

## Environment (please complete the following information):
 - Build name 0.0.20210410.1-2104.2-prod
 - Database version 9.2.21035.00163
 - PackageDeployerTemplate 0.3.6

**Comment by tdashworth** [comment]
After more investigation, I found that including a custom connector in a solution creates it as new within the environment. This means a new `connectorinternalid` between source and target environments. The source `connectorinternalid` is referenced in the connection reference in the `connectorid` column. which needs to be replaced with the created one in target environment. 

To find this, the connection reference has a lookup to `connector` called `customconnectorid` which uses a different id that is the same across environments. Once you have the connector, the `connectorid` can be updated on the connection reference using the `connectorinternalid` on the connector. 

I will post a diagram to visually explain 😄

**Comment by ewingjm** [comment]
Is it worth categorising this as an enhancement? I'd say it's more likely a bug in the docs that we don't state it doesn't support scenarios where the connection reference logical names aren't known at deploy time - but it's not an issue with the current functionality to map connections to connection references by logical names.

**Comment by tdashworth** [comment]
![image](https://user-images.githubusercontent.com/9384593/115676789-d4047700-a347-11eb-853a-c84c3b213c26.png)

**Comment by ewingjm** [comment]
Ah, perhaps I misunderstood the issue, then. So is this actually a defect in how Microsoft handle importing solutions with custom connectors and connection references?

**Comment by tdashworth** [comment]
Yeah, I would say so. The custom connector effectively changes name and therefore "Not Found" in the target environment.

**Comment by tdashworth** [comment]
I have tried importing a solution manually, where the Connection References pane is shown, and I am unable to create a connection at this stage: 

![image](https://user-images.githubusercontent.com/9384593/115772704-55d7bd00-a3a7-11eb-967b-fece7678cb2e.png)

![image](https://user-images.githubusercontent.com/9384593/115772456-0a251380-a3a7-11eb-94f7-4001b589e13b.png)

This appears to entirely block this manual process as setting a connection is a required step.

**Comment by ewingjm** [comment]
So this seems like quite a high severity Microsoft defect. You basically can't ship solutions that use flows with custom connectors?

**Comment by tdashworth** [comment]
Correct. I will raise this with them first thing in the morning.

**Comment by tdashworth** [comment]
Some further testing results...

I first created some solutions: 
- Solution 1:
  - Custom Connector
- Solution 2:
  - Connection Reference (to Custom Connector)
- Solution 1 & 2:
  - Custom Connector (same component in Solution 1)
  - Connection Reference (same component in Solution 2) 

I also created a Connection for the Custom Connector. 

Next, I deleted the Connector and the Reference but left the Connection. 

My first test was to install Solution 1 (Managed) then Solution 2 (Managed) which worked and required me to create a new Connection for the Reference despite a connection already existing. I then uninstalled both managed solutions.

My next test was to install Solution 1 & 2 (Managed) which again asked for a Connection for the Reference although this failed as per my previous comment and the solution was un-installable via the Maker Portal. 

My last test was to install Solution 1 (Managed) then Solution 1 & 2 (Managed) which again worked similar to my first test. 

---

In conclusion, the Custom Connector must be in a different solution and pre-installed within the environment to be able to import solutions with a Connection Reference to it.

**Comment by tdashworth** [comment]
So I've been running more tests will the package deployer and it seems to work today without any changes... 

I first ran the package deployer without any custom connection reference which successes.
Then I created a connection for the custom connector (Maker Portal > Data > Custom Connections > + ).
Next, I set the connection reference via an environment variable and ran the deployer again.

**Comment by tdashworth** [comment]
Solution Files used for testing:
[CustomConnectionTest1_1_0_0_2_unmanaged.zip](https://github.com/Capgemini/powerapps-packagedeployer-template/files/6384132/CustomConnectionTest1_1_0_0_2_unmanaged.zip)
[CustomConnectionTest1_1_0_0_1_managed.zip](https://github.com/Capgemini/powerapps-packagedeployer-template/files/6384133/CustomConnectionTest1_1_0_0_1_managed.zip)
[CustomConnectionTest2_1_0_0_1_managed.zip](https://github.com/Capgemini/powerapps-packagedeployer-template/files/6384131/CustomConnectionTest2_1_0_0_1_managed.zip)
[CustomConnection12_1_0_0_1_managed.zip](https://github.com/Capgemini/powerapps-packagedeployer-template/files/6384134/CustomConnection12_1_0_0_1_managed.zip)

**Comment by tdashworth** [comment]
I have manually tested this code (updating the `connectorid` on the `connectionreference` then redeployed so this fix should work. But I'm not 100% it works for all environments.

**Comment by tdashworth** [comment]
So after writing the fix and merging, it doesn't appear this is a problem anymore so I'm closing this issue.
